api/agent/agent.go

```diff
@@ -567,16 +596,28 @@ func (a *agent) prepCold(ctx context.Context, call *call, tok ResourceToken, ch
 
 	// pull & create container before we return a slot, so as to be friendly
 	// about timing out if this takes a while...
-	cookie, err := a.driver.Prepare(ctx, container)
-	slot := &coldSlot{cookie, tok, err}
+	var err error
+	var cookie drivers.Cookie
+	var waiter drivers.WaitResult
+
+	cookie, err = a.driver.Prepare(ctx, container)
+	if err == nil {
+		waiter, err = cookie.Run(ctx)
```

Comment:
the issue here is changing the ordering of this means that the call will be running before we have called `call.Start` -- which we want to use to manage the task state (specifically, for async, but also sync has some field updates). I'm not sure it's worth trying to dangle this down here, trying to keep it so that `submit` is 'prepare' -> 'start' -> 'exec' -> 'end' and running the container for cold should remain in exec imo. i could elaborate on finer details, but the timings are off for a few things (the task finishing before the start time is set seems... odd). in any case, it doesn't seem worth moving this for state management alone and we'll have to make a number of other adjustments.

wrt state management, I'm less sure of the implications. my mental model is not yet built, but it seems like we're kind of lying here anyway saying the container is idle after calling run. we should set to execution before calling cookie run I think (and this should be in exec where that happens, I think). at least, moving this back to its old location seems sane, and for cold there really is no 'idle' container transition, there is basically only... created but not yet running (not really idle, but kind of I guess), not sure if we'd like to skip it [for cold] or if we're relying on certain state transitions for correctness somewhere, but leaving the state transition to idle here and moving cookie run back also seems to make sense to me.

Reply:
I agree, this was a tricky move... Start() interaction, etc. I'll revert this one...